cobc/settings.c

```diff
@@ -0,0 +1,446 @@
+/*
+   Copyright (C) 2001-2023 Free Software Foundation, Inc.
+
+   Authors:
+   Keisuke Nishida, Roger While, Ron Norman, Simon Sobisch, Brian Tiffin,
+   Edward Hart, Dave Pitts, Fabrice Le Fessant
+
+   This file is part of GnuCOBOL.
+
+   The GnuCOBOL compiler is free software: you can redistribute it
+   and/or modify it under the terms of the GNU General Public License
+   as published by the Free Software Foundation, either version 3 of the
+   License, or (at your option) any later version.
+
+   GnuCOBOL is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with GnuCOBOL.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "tarstamp.h"
+#include "config.h"
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#ifdef	HAVE_UNISTD_H
+#include <unistd.h>
+#endif
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <stddef.h>
+#include <stdarg.h>
+#include <string.h>
+#ifdef	HAVE_STRINGS_H
+#include <strings.h>
+#endif
+#include <ctype.h>
+#include <time.h>
+#include <limits.h>
+
+#include "cobc.h"
+#include "tree.h"
+
+/*
+Settings are used to override GnuCOBOL host config, i.e. parameters
+that are used to adapt GnuCOBOL to the system running it. By default,
+these parameters are guessed at compilation time. However, it happens
+that we want to change them at runtime:
+
+- For relocation, i.e. to install GnuCOBOL in a different directory
+  from the one in which it was supposed to be.
+
+- For retargetting, i.e. to target a different C compiler or linker
+  that the one that was detected at compile time.
+
+Use `gnucobol --settings` to output the current settings. The output
+can be copied into a file, that will later be read by GnuCOBOL:
+
+```
+$ gnucobol --settings > gnucobol.settings
+$ gnucobol --settings=gnucobol.settings
+```
+
+You can then modify any option within the file.
+
+A file containing these settings can be specified in 3 different ways,
+the first working way in this order will be used:
+
+* Use of the argument `--settings=FILE`
+* Use of the environment variable `COBC_SETTINGS`
+* Use of `../etc/gnucobol/gnucobol.settings` from the directory
+  containing the `cobc` executable
+
+If a file is specified, an error while loading the file is fatal.
+
+Once the settings have been loaded, a few variables can be modified if
```

Comment:
I suggest to make that not depending on `COBC_IS_RELOCATABLE` (env var or its default) but instead just checking if the executable is in the place that we expect - if it isn't then o the relocation.
... but this part should possibly go elsewhere (see main review comment)